### ISSUE TITLE
Support media types with a suffix

### ISSUE THREAD

**Issue by delitescere** [juxt/yada#188]
Media types may include a suffix to indicate the underlying structure of the content. A well-known example is `application/hal+json` which indicates a HAL document in JSON format.

Where the media type of the suffix (e.g. `application/json`) already has a render function (`render-map`, etc), an additional render function should not be needed for the suffixed media type (e.g. `application/hal+json`), unless specific alternate processing is desired.

This means that if there is a `render-map` function defined for `application/json` then a `render-map` for `application/hal+json` need not be defined (but may be if desired). Yada should default to the render functions for the media type of the suffix if there are none found matching the suffixed media type.

Nonetheless, in the `:produces` declarations for a resource, the specific media types should still be explicitly listed if they are to be accepted based on `Accept` request header content negotiation and used as the `Content-Type` response header. For example:

```clojure
;; ...
:produces ["text/html"
           "application/json-home"
           "application/hal+json"
           "application/json"
           "application/edn"]
;; ...
```

That it is, even though the `application/hal+json` representation _is valid JSON_, if it wasn't the negotiated content type, then it should not be nominated as the content type in the response. Indeed, some providers may wish to elide the HAL metadata from the response entity if HAL wasn't requested (even though, again, a consumer desiring a JSON response will receive valid JSON even with the HAL metadata included).

Further reading:
https://en.wikipedia.org/wiki/Media_type#Suffix
https://tools.ietf.org/html/rfc6839

**Comment by malcolmsparks** [comment]
Completely agree. Should be simple to implement. Happy to do so or accept a PR. The defmulti needs to call `clojure.core/methods` on itself and fallback to the non-suffix version.

**Comment by delitescere** [comment]
I'll take a crack at it with that hint. Cheers!


----------------------------------
From the phone

**Comment by malcolmsparks** [comment]
If you haven't already figured it out, I've added a note to the README about how to build and install your own version of yada:

https://github.com/juxt/yada#contributing

**Comment by SevereOverfl0w** [comment]
https://trac.tools.ietf.org/html/draft-ietf-appsawg-media-type-suffix-regs-02#section-3.5

We should be cautious of examples like this one when deciding how to implement suffixes. You cannot simply go from from `application/foo+wbxml` to `application/wbxml` as the correct ancestor would be `application/vnd.wap.wbxml`.

I would assume this means any solution must be solved with a registry.

I imagine it's also possible to create a content-type like `application/bar+html`, but I can't see any example in the registry where I believe that would be the case.

**Comment by delitescere** [comment]
Good point, but I’m not sure it need be as complex as a registry.

Taking a view of convention-over-configuration, and the view that your example has a poorly- and/or incorrectly-chosen suffix, and that the service provider decidedly knows which media types it supports: the framework can do the conventional thing (simple de-concatenation per the spec), and some augmented configuration could deal with the edge cases for a specific application.

**Comment by malcolmsparks** [comment]
This is implemented now in 1.3.0-alpha10, but only for response representations, not for request bodies. That said, this feels like 95% of use-cases so worth closing this issue. Feel free to raise a new issue for request bodies if you have a need.
